src/logics/ArithLogic.cc

```diff
@@ -243,35 +245,196 @@ PTRef ArithLogic::normalizeMul(PTRef mul)
+                    constantPoly.addTerm(PTRef_Undef, coeff * logic.getNumConst(val));
+                    poly.merge(constantPoly, 1);
+                }
+                changed = true;
+            }
+        }
+        // remove used polynomials
+        auto lastValid = zeroPolynomials.size() - 1;
+        for (auto rit = processedIndices.rbegin(); rit != processedIndices.rend(); ++rit) {
+            auto index = *rit;
+            zeroPolynomials[index] = std::move(zeroPolynomials[lastValid]);
+            zeroPolynomials.pop_back();
+            --lastValid;
+        }
+
+        if (not changed) { break; }
+    }
+    return substitutions;
+}
+
+PTRef polyToPTRef(ArithLogic & logic, poly_t const & poly) {
+    vec<PTRef> args;
+    args.capacity(poly.size());
+    assert(poly.size() > 0);
+    assert(poly.begin()->var != PTRef_Undef);
+    SRef sortRef = logic.getSortRef(poly.begin()->var);
+    for (auto const & term : poly) {
+        assert(not term.coeff.isZero());
+        if (term.var == PTRef_Undef) { args.push(logic.mkConst(sortRef, term.coeff)); }
+        else if (term.coeff.isOne()) { args.push(term.var); }
+        else { args.push(logic.mkTimes(term.var, logic.mkConst(logic.getSortRef(term.var), term.coeff))); }
+    }
+    return logic.mkPlus(std::move(args)); // TODO: Can we use non-simplifying versions of mkPlus/mkTimes?
+}
+
+Logic::SubstMap collectSingleEqualitySubstitutions(ArithLogic & logic, std::vector<poly_t> & zeroPolynomials) {
+    Logic::SubstMap substitutions;
+    std::unordered_map<PTRef, std::vector<std::size_t>, PTRefHash> varToPolyIndices;
+
+    for (std::size_t i = 0; i < zeroPolynomials.size(); ++i) {
+        auto const & poly = zeroPolynomials[i];
+        for (auto const & term : poly) {
+            varToPolyIndices[term.var].push_back(i);
+        }
+    }
+
+    std::unordered_set<std::size_t> processedIndices;
+    for (auto const & [var, polyIndices] : varToPolyIndices) {
+        if (polyIndices.size() != 1 or var == PTRef_Undef) { continue; }
+        auto index = polyIndices[0];
+        auto [it, inserted] = processedIndices.insert(index);
+        if (not inserted) { continue; }
+        auto & poly = zeroPolynomials[index];
+        auto coeff = poly.removeVar(var);
+        coeff.negate();
+        if (not coeff.isOne()) {
+            poly.divideBy(coeff);
+        }
+        PTRef val = polyToPTRef(logic, poly);
+        substitutions.insert(var, val);
+    }
+    // Remove processed polynomials
+    std::vector<std::size_t> indicesToRemove(processedIndices.begin(), processedIndices.end());
+    std::sort(indicesToRemove.begin(), indicesToRemove.end());
+    auto lastValid = zeroPolynomials.size() - 1;
+    for (auto rit = indicesToRemove.rbegin(); rit != indicesToRemove.rend(); ++rit) {
+        auto index = *rit;
+        zeroPolynomials[index] = std::move(zeroPolynomials[lastValid]);
+        zeroPolynomials.pop_back();
+        --lastValid;
```

Comment:
```suggestion
    for (auto rit = indicesToRemove.rbegin(); rit != indicesToRemove.rend(); ++rit) {
        zeroPolynomials[*rit] = std::move(zeroPolynomials.back());
        zeroPolynomials.pop_back();
```

Reply:
Yes, I wanted to have the `.back()` version.
I'll include these.

Reply:
Pushed new version.